lib/id3c/cli/command/etl/redcap_det.py

```diff
@@ -55,12 +56,34 @@ def get_redcap_record(record_id: str, token_name: str) -> dict:
         'Accept': 'application/json'
     }
 
-    response = requests.post(os.environ['REDCAP_API_URL'], data=data, headers=headers)
+    response = requests.post(url, data=data, headers=headers)
     response.raise_for_status()
 
     return response.json()[0]
 
 
+def get_redcap_api_credentials() -> Tuple[str, str]:
+    """
+    Returns a tuple of ``(url, token)`` for use with REDCap's web API.
+
+    Requires the environmental variables ``REDCAP_API_URL`` and
+    ``REDCAP_API_TOKEN``.
+    """
+    url = os.environ.get("REDCAP_API_URL")
+    token = os.environ.get("REDCAP_API_TOKEN")
```

Comment:
How does this function allow other ways than the environment to provide values when the environment variable appears to be hard-coded here as `REDCAP_API_TOKEN`? 

Each REDCap project requires a different API token, and I don't see why this revision is preferred.

Reply:
Would the `REDCAP_API_TOKEN` need to be declared before running every etl? 

e.g.
```
REDCAP_API_TOKEN=$REDCAP_SINGLE_SWAB_AND_SEND_API_TOKEN PGDATABASE=seattleflu pipenv run id3c etl redcap-det swab-n-send
```

Reply:
> How does this function allow other ways than the environment to provide values when the environment variable appears to be hard-coded…?

It does not allow other ways at the moment, but using a function as a credentials provider means it is easier to swap out later without changing calling code or repeating environment variable names in many places.

> Each REDCap project requires a different API token, and I don't see why this revision is preferred.
> Would the REDCAP_API_TOKEN need to be declared before running every etl?

Parameterizing the environment variable name on the REDCap project feels like an antipattern to me. A consistent environment variable for specifying the token makes it predictable and less to document and remember. Rather than setting all environment variables at once, I expect that each project's will be set in separate invocations (like your example) or with separate environment files or envdirs for each project. For example, the crontab we'll end up with might look like:

```
REDCAP_API_URL=https://redcap.iths.org

0 * * * * REDCAP_API_TOKEN=deadbeef pipenv run id3c etl redcap-det kiosk
5 * * * * REDCAP_API_TOKEN=abadcafe pipenv run id3c etl redcap-det swab-n-send
…
```

or

```
0 * * * * envdir /opt/secrets/redcap-kiosk/ pipenv run id3c etl redcap-det kiosk
5 * * * * envdir /opt/secrets/redcap-swab-n-send/ pipenv run id3c etl redcap-det swab-n-send
…
```

or

```
0 * * * * source /opt/secrets/redcap-kiosk.env && pipenv run id3c etl redcap-det kiosk
5 * * * * source /opt/secrets/redcap-swab-n-send.env && pipenv run id3c etl redcap-det swab-n-send
…
```

Environment variables are global variables (albeit special ones), and just like with global variables, it's good to minimize both the number of different ones and the number of places they're accessed in the code.

Reply:
Thanks for the examples! You have convinced me.